pandas/tests/indexes/multi/test_names.py

```diff
@@ -148,3 +148,31 @@ def test_setting_names_from_levels_raises():
     assert pd.Index._no_setting_name is False
     assert pd.Int64Index._no_setting_name is False
     assert pd.RangeIndex._no_setting_name is False
+
+
+@pytest.mark.parametrize("func", ["rename", "set_names"])
+@pytest.mark.parametrize(
+    "rename_dict, exp_names",
+    [
+        ({"x": "z"}, ["z", "y", "z"]),
+        ({"x": "z", "y": "x"}, ["z", "x", "z"]),
+        ({"y": "z"}, ["x", "z", "x"]),
+        ({}, ["x", "y", "x"]),
+        ({"z": "a"}, ["x", "y", "x"]),
+        ({"y": "z", "a": "b"}, ["x", "z", "x"]),
+    ],
+)
+def test_name_mi_with_dict_like(func, rename_dict, exp_names):
+    # GH#20421
+    mi = MultiIndex.from_arrays([[1, 2], [3, 4], [5, 6]], names=["x", "y", "x"])
```

Comment:
Can you also include a test for the use case 
```
mi = MultiIndex.from_arrays([[1, 2], [3, 4], [5, 6]], names=["x", "y", "z"])
result = mi.rename({"x":"a"})
```
In other words, start with a `MultiIndex` that has different names for the level, and make sure they rename correctly.

We also should decide if we allow duplicate level names (which your test is producing).  But maybe that's a separate issue.

Reply:
Added the tests.